fold/fold.go

```diff
@@ -0,0 +1,895 @@
+	}
+
+	// if there's a helix, start,end counts as well
+	if helix {
+		branches = append(branches, Subsequence{start, end})
+	}
+
+	// count up unpaired bp and asymmetry
+	branches_count := len(branches)
+	unpaired := 0
+	e_sum := 0.0
+	ij := Subsequence{start, end}
+	for index, ij2 := range branches {
+		i2, j2 := ij2.Start, ij2.End
+		ij1 := branches[abs((index-1)%len(branches))]
+		j1 := ij1.End
+		ij3 := branches[abs((index+1)%len(branches))]
+		i3, j3 := ij3.Start, ij3.End
+
+		// add energy from unpaired bp to the right
+		// of the helix as though it was a dangling end
+		// if there's only one bp, it goes to whichever
+		// helix (this or the next) has the more favorable energy
+		unpaired_left := 0
+		unpaired_right := 0
+		de := 0.0
+		if index == len(branches)-1 && !helix {
+			// pass
+		} else if ij3 == ij {
+			unpaired_left = i2 - j1 - 1
+			unpaired_right = j3 - j2 - 1
+
+			if unpaired_left != 0 && unpaired_right != 0 {
+				de = Stack(i2-1, i2, j2+1, j2, foldContext)
+			} else if unpaired_right != 0 {
+				de = Stack(-1, i2, j2+1, j2, foldContext)
+				if unpaired_right == 1 {
+					de = min(Stack(i3, -1, j3, j3-1, foldContext), de)
+				}
+			}
+		} else if ij2 == ij {
+			unpaired_left = j2 - j1 - 1
+			unpaired_right = i3 - i2 - 1
+
+			if unpaired_left != 0 && unpaired_right != 0 {
+				de = Stack(i2-1, i2, j2+1, j2, foldContext)
+			} else if unpaired_right != 0 {
+				de = Stack(i2, i2+1, j2, -1, foldContext)
+				if unpaired_right == 1 {
+					de = min(Stack(i3-1, i3, -1, j3, foldContext), de)
+				}
+			}
+		} else {
+			unpaired_left = i2 - j1 - 1
+			unpaired_right = i3 - j2 - 1
+
+			if unpaired_left != 0 && unpaired_right != 0 {
+				de = Stack(i2-1, i2, j2+1, j2, foldContext)
+			} else if unpaired_right != 0 {
+				de = Stack(-1, i2, j2+1, j2, foldContext)
+				if unpaired_right == 1 {
+					de = min(Stack(i2-1, i2, j2+1, j2, foldContext), de)
+				}
+			}
+		}
+		e_sum += de
+		unpaired += unpaired_right
+		if unpaired_right < 0 {
+			return DefaultStructure, fmt.Errorf("Multibranch: subsequence (%d, %d, %d): unpaired_right < 0", start, end, k)
+		}
+
+		if ij2 != ij { // add energy
+			w, err := W(i2, j2, foldContext)
+			if err != nil {
+				return DefaultStructure, fmt.Errorf("Multibranch: subsequence (%d, %d, %d): %w", start, end, k, err)
+			}
+			e_sum += w.Energy
+		}
+
+	}
+
+	if unpaired < 0 {
+		return DefaultStructure, fmt.Errorf("Multibranch: subsequence (%d, %d, %d): unpaired < 0", start, end, k)
+	}
+
+	// penalty for unmatched bp and multi-branch
+	a, b, c, d := foldContext.Energies.Multibranch.A, foldContext.Energies.Multibranch.B, foldContext.Energies.Multibranch.C, foldContext.Energies.Multibranch.D
+	e_Multibranch := a + b*float64(len(branches)) + c*float64(unpaired)
+
+	if unpaired == 0 {
+		e_Multibranch = a + d
+	}
+
+	// energy of min-energy neighbors
+	e := e_Multibranch + e_sum
+
+	// pointer to next structures
+	if helix {
+		// branches.pop()
+		branches = branches[:len(branches)-1]
```

Comment:
get the feeling this should be optimized?

Reply:
I'm not sure what you mean